crates/net/network/src/fetch/mod.rs

```diff
@@ -232,16 +242,31 @@ impl StateFetcher {
             ))
         }
 
-        if let Some(peer) = self.peers.get_mut(&peer_id) {
-            // If the peer is still ready to be accept new requests, we try to send a followup
-            // request immediately.
-            if peer.state.on_request_finished() {
-                return self.followup_request(peer_id)
+        if !is_likely_a_bad_message {
+            if let Some(peer) = self.peers.get_mut(&peer_id) {
+                // If the peer is still ready to be accept new requests, we try to send a followup
+                // request immediately.
+                if peer.state.on_request_finished() {
+                    return self.followup_request(peer_id)
+                }
             }
         }
+
         None
     }
 
+    fn is_likely_bad_message(request: &HeadersRequest, headers: &[Header]) -> bool {
+        if request.limit != headers.len() as u64 {
```

Comment:
limit is only the upper bound of the requested range, the peer may return less then requested, which is valid.

But what usually happens is that peers just return one or none headers, so we should check for `< 1  && request.limit != len`